rfd/0113-moderated-file-transfers.md

```diff
@@ -0,0 +1,176 @@
+
+After the steps above we are in the `PENDING` state. This isn't a real "state" that exists in any struct, just the part of the flow that is waiting for responses. This is not a blocking state and essentially just creates the request, and updates the relevant UI (more on the UI below). 
+
+#### The Approval process
+
+Similar to sending the `FileTransferRequest` event, another event will be listened to, the `FileTranferRequestResponse` (working name). We could separate this new event into two with Approve/Deny but a simple switch case on a field should suffice. 
+
+OnDeny:
+	1. Remove request from the session struct. No need to keep it around in a "Denied" state, just delete.
+	2. Emit audit event "XUser has denied ~/myfile.txt download request from michael".
+	3. Broadcast message same as audit event above so the rest of the session can see
+	4. Notify all members of the denial so the UI can be removed.
+
+OnApprove:
+	1. Add approver to the `approvers` slice
+	2. Broadcast/audit event
+	3. We can then use a policy checker to see if the approvers fulfill any moderation policy on the original requester. We can treat this check the same as the `checkIfStart` conditional for opening a session. If this comes back true, we notify the original requester with an event containing the ID of the `FileTransferRequest`
+
+Once the client receives this final "approved" message, we can automatically send a "normal" SCP request (over HTTP) with two new optional params, `sessionID` and `requestID` (similar to the new optional `webauthn` param in this same request). The benefits of using the normal SCP request is that we can conditionally choose to skip this entire approval process flow for non-moderated sessions. If the session is not moderated, just send the scp request as usual. If it is, do the song and dance perscribed above.
+
+#### Updated file transfer api handler
+
+Not much will change here. The only difference is adding the `requestID` and `sessionID` as env vars in the `serverContext`. This will allow us to extend our `SessionRegistry.OpenExecSession` method with a new check like `isApprovedRequest`
+
+```go
+func (s *SessionRegistry) isApprovedRequest(ctx context.Context, scx *ServerContext) (bool error) {
+	// fetch session from registry with sessionID
+
+	// find request in the session by requestID
+
+	// validate incoming request (user/file/direction/etc) against the stored request
+
+	// if all is good, TRUE
+}
+```
+
+Then our `OpenExecSession` can be updated:
+
+```go
+func (s *SessionRegistry) OpenExecSession(ctx context.Context, channel ssh.Channel, scx *ServerContext) error {
+	// ...
+	if !canStart && !isApprovedRequest {
+		return errCannotStartUnattendedSession
+	}
+
+	// ...
+}
+```
+
+#### UX/UI
+
+The request-a-request flow would be: the request -> the approval -> the file transfer
+
+The `ApprovedFileTransferRequest` will be stored in-memory and not need to be persisted as it's own entity anywhere else.
```

Comment:
It does need to propagate to all proxies though, right?

Reply:
The `ApprovedFileTransferRequest` struct only needs to live in the node memory and doesn't need to be propagated to the proxies. the proxy in this case is only used to send the request/approval through the ssh channel connected to the proxy for each member.